File: game_of_life.rb
class Grid
  attr_reader :x, :y
  def initialize(x, y)
    @x, @y = x, y
  end
end

class Cell
  attr_reader :w, :h
  def initialize(w, h)
    @w, @h = w, h
  end
end

class GameOfLife
  def alert value
    `alert(value)`
  end
  
  def initialize(canvas, seed)
    @canvas =  nil     # CanvasRenderingContext2D object (We store it like this because its still possible to get the dom node out of it, which we use less)
    @gen =  0          # Generation Counter
    @genc =  []        # Current Generation
    @genn =  []        # Next Generation
    @genlog =  []      # A JSON log of the last # generations to detect stability. JSON is used because its easier to compare than an two entire arrays, and its standard.
    @history =  5      # How many generations to log (ie what size of oscillating pattern to detect)
    @seedweight =  0.2 # 0 to 1 of density of life:death for initial random seed. 1 is complete life, 0 complete death. High densities are pointless, most cells die on second generation.
    @grid = Grid.new(50, 50)          # Size of the grid, in cells
    @cell = Cell.new(10, 10)          # Size of cells, in pixels
    
    # if ! canvas.canvas? || ! canvas.nodeName == "CANVAS" # We've either been given a CanvasRenderingContext2D, which has a .canvas property, or we've been given a canvas DOM node, which will have a nodeName
    # raise "Game needs a canvas object or canvas DOM node"
    # @canvas = canvas.nodeName? ? canvas.getContext("2d") : canvas
    @canvas = canvas
    if !seed
      # generate initial seed.
      @genc = (0..@grid.x).to_a.map { |x|
        (0..@grid.y).to_a.map { |y|
          if @seedweight > `Math.random()`
            1
          else
            0
          end
        }
      }
    else
      @genc = seed
      @genn = @genc
    end
    #    end            
  end
  
  def _countNeighbours(x, y)
    xp = x + 1 > @grid.x - 1 ? 0 : x + 1
    xm = x - 1 < 0 ? @grid.x - 1 : x - 1
    yp = y + 1 > @grid.y - 1 ? 0 : y + 1
    ym = y - 1 < 0 ? @grid.y - 1 : y - 1
    # {top left} {top} {top right} {left} {right} {bottom left} {bottom} {bottom right}
    sum = @genc[xm][ym] + @genc[x][ym] + @genc[xp][ym] + @genc[xm][y] + @genc[xp][y] + @genc[xm][yp] + @genc[x][yp] + @genc[xp][yp]
    #puts "sum = #{sum}"
    sum
  end

  def _getState(x, y)
    cell = @genc[x][y]
    pop = _countNeighbours(x, y)
    # Standard rules for survival
    return 0 if cell == 1 and pop < 2 
    return 1 if cell == 1 and ((2 == pop) or (pop == 3) )
    return 0 if cell == 1  and pop > 3 
    return 1 if cell == 0 and pop == 3 
    0
  end

  # Iterate over the next array and get the new state for each cell
  def runGeneration
    @genn = (0..@grid.x).to_a.map { |x|
      (0..@grid.y).to_a.map { |y|
        _getState(x, y)
      }
    }
    _rotateGenerations()
    @genc = @genn
    @gen += 1
  end
	# Rotate the history log by popping off the old and unshifting in the new
  def _rotateGenerations
    @genlog.pop() if @genlog.length > @history
    @genlog.unshift(`JSON.stringify(#@genn)`)
    puts @genlog
  end
	# Check for stability in the last @history generations by matching the log against the current generation.
  def checkStable
    json_genc = `JSON.stringify(#@genc)`
     @genlog.each { |key, gen|
      return true if gen == json_genc and key > 0
    }
    return false
  end
  # Make sure our canvas is appropriately sized, also used for clearing a canvas pre render
  def _sizeCanvas
    # Use the attributes because using the css means the canvas
    # stretches.
    `#@canvas.canvas.width = #{ @grid.x * @cell.w}`
    `#@canvas.canvas.height = #{@grid.y * @cell.h} `
    return true
  end
  # Render the current generation the the canvas element

  def render
    _sizeCanvas()
    (0..@grid.x).to_a.each { |x|
      (0..@grid.y).to_a.each { |y|
        `#@canvas.fillStyle = #@genc[x][y] ? "black" : "white"`
        `#@canvas.fillRect(x * #@cell.w, y * #@cell.h, #@cell.w, #@cell.h)`
      }
    }
    return true
  end
  
end


class Interval
  def initialize(time=0, &block)
    @interval = `setInterval(function(){#{block.call}}, time)`
  end

  def stop
    interval = @interval
    `clearInterval(interval)`
  end
end

class GameRunner
  def initialize( interval)
    @game = GameOfLife.new(`document.getElementById('board').getContext("2d")`)
    myself = self
    @interval = Interval.new interval do
      myself.work
    end
    @game.render
  end

  def work
    @game.runGeneration()
    @game.render()
    `document.getElementById('info').innerHTML = "Generation: " + #@game.gen`
    @interval.stop if @game.checkStable() 
  end

end

do_it = GameRunner.new 10

File: game_of_life.js
(function() {
  var __opal = Opal, self = __opal.top, __scope = __opal, nil = __opal.nil, __breaker = __opal.breaker, __slice = __opal.slice, __klass = __opal.klass, __range = __opal.range;
  var game = nil, do_it = nil;
  (function(__base, __super){
    // line 1, (file), class Grid
    function Grid() {};
    Grid = __klass(__base, __super, "Grid", Grid);
    var Grid_prototype = Grid.prototype, __scope = Grid._scope;
    Grid_prototype.x = Grid_prototype.y = nil;

    // line 2, (file), Grid#x
    Grid_prototype.$x = function() {
      
      return this.x
    }, 
    // line 2, (file), Grid#y
    Grid_prototype.$y = function() {
      
      return this.y
    };

    // line 3, (file), Grid#initialize
    Grid_prototype.$initialize = function(x, y) {
      var __a;
      return __a = [x, y], this.x = __a[0], this.y = __a[1];
    };

  })(self, null);
  (function(__base, __super){
    // line 8, (file), class Cell
    function Cell() {};
    Cell = __klass(__base, __super, "Cell", Cell);
    var Cell_prototype = Cell.prototype, __scope = Cell._scope;
    Cell_prototype.w = Cell_prototype.h = nil;

    // line 9, (file), Cell#w
    Cell_prototype.$w = function() {
      
      return this.w
    }, 
    // line 9, (file), Cell#h
    Cell_prototype.$h = function() {
      
      return this.h
    };

    // line 10, (file), Cell#initialize
    Cell_prototype.$initialize = function(w, h) {
      var __a;
      return __a = [w, h], this.w = __a[0], this.h = __a[1];
    };

  })(self, null);
  (function(__base, __super){
    // line 15, (file), class GameOfLife
    function GameOfLife() {};
    GameOfLife = __klass(__base, __super, "GameOfLife", GameOfLife);
    var GameOfLife_prototype = GameOfLife.prototype, __scope = GameOfLife._scope;
    GameOfLife_prototype.grid = GameOfLife_prototype.genc = GameOfLife_prototype.genn = GameOfLife_prototype.gen = GameOfLife_prototype.genlog = GameOfLife_prototype.history = GameOfLife_prototype.canvas = GameOfLife_prototype.cell = nil;

    // line 16, (file), GameOfLife#alert
    GameOfLife_prototype.$alert = function(value) {
      
      return alert(value);
    };

    // line 20, (file), GameOfLife#initialize
    GameOfLife_prototype.$initialize = function(canvas, seed) {
      var __a, __b;
      this.canvas = nil;
      this.gen = 0;
      this.genc = [];
      this.genn = [];
      this.genlog = [];
      this.history = 5;
      this.seedweight = 0.2;
      this.grid = __scope.Grid.$new(50, 50);
      this.cell = __scope.Cell.$new(10, 10);
      this.canvas = canvas;
      if ((__a = !seed) !== false && __a !== nil) {
        return this.genc = (__b = __range(0, this.grid.$x(), false).$to_a(), __b.$map._p = (__a = function(x) {

          var __a, __b;
          if (this.grid == null) this.grid = nil;

          if (x == null) x = nil;

          return (__b = __range(0, this.grid.$y(), false).$to_a(), __b.$map._p = (__a = function(y) {

            
            if (this.seedweight == null) this.seedweight = nil;

            if (y == null) y = nil;

            if (this.seedweight['$>'](Math.random())) {
              return 1
              } else {
              return 0
            }
          }, __a._s = this, __a), __b.$map())
        }, __a._s = this, __a), __b.$map())
        } else {
        this.genc = seed;
        return this.genn = this.genc;
      };
    };

    // line 53, (file), GameOfLife#_countNeighbours
    GameOfLife_prototype.$_countNeighbours = function(x, y) {
      var xp = nil, xm = nil, yp = nil, ym = nil, sum = nil, __a, __b, __c, __d, __e, __f, __g, __h, __i, __j, __k, __l, __m, __n;
      xp = (function() { if ((__a = x, __b = 1, typeof(__a) === 'number' ? __a + __b : __a['$+'](__b))['$>']((__a = this.grid.$x(), __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b)))) {
        return 0
        } else {
        return (__a = x, __b = 1, typeof(__a) === 'number' ? __a + __b : __a['$+'](__b))
      }; return nil; }).call(this);
      xm = (function() { if ((__a = x, __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b))['$<'](0)) {
        return (__a = this.grid.$x(), __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b))
        } else {
        return (__a = x, __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b))
      }; return nil; }).call(this);
      yp = (function() { if ((__a = y, __b = 1, typeof(__a) === 'number' ? __a + __b : __a['$+'](__b))['$>']((__a = this.grid.$y(), __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b)))) {
        return 0
        } else {
        return (__a = y, __b = 1, typeof(__a) === 'number' ? __a + __b : __a['$+'](__b))
      }; return nil; }).call(this);
      ym = (function() { if ((__a = y, __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b))['$<'](0)) {
        return (__a = this.grid.$y(), __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b))
        } else {
        return (__a = y, __b = 1, typeof(__a) === 'number' ? __a - __b : __a['$-'](__b))
      }; return nil; }).call(this);
      sum = (__a = (__c = (__e = (__g = (__i = (__k = (__m = this.genc['$[]'](xm)['$[]'](ym), __n = this.genc['$[]'](x)['$[]'](ym), typeof(__m) === 'number' ? __m + __n : __m['$+'](__n)), __l = this.genc['$[]'](xp)['$[]'](ym), typeof(__k) === 'number' ? __k + __l : __k['$+'](__l)), __j = this.genc['$[]'](xm)['$[]'](y), typeof(__i) === 'number' ? __i + __j : __i['$+'](__j)), __h = this.genc['$[]'](xp)['$[]'](y), typeof(__g) === 'number' ? __g + __h : __g['$+'](__h)), __f = this.genc['$[]'](xm)['$[]'](yp), typeof(__e) === 'number' ? __e + __f : __e['$+'](__f)), __d = this.genc['$[]'](x)['$[]'](yp), typeof(__c) === 'number' ? __c + __d : __c['$+'](__d)), __b = this.genc['$[]'](xp)['$[]'](yp), typeof(__a) === 'number' ? __a + __b : __a['$+'](__b));
      return sum;
    };

    // line 64, (file), GameOfLife#_getState
    GameOfLife_prototype.$_getState = function(x, y) {
      var cell = nil, pop = nil, __a, __b, __c;
      cell = this.genc['$[]'](x)['$[]'](y);
      pop = this.$_countNeighbours(x, y);
      if ((__a = ((__b = cell['$=='](1)) ? pop['$<'](2) : __b)) !== false && __a !== nil) {
        return 0
      };
      if ((__a = ((__b = cell['$=='](1)) ? ((__c = (2)['$=='](pop)), __c !== false && __c !== nil ? __c : pop['$=='](3)) : __b)) !== false && __a !== nil) {
        return 1
      };
      if ((__a = ((__b = cell['$=='](1)) ? pop['$>'](3) : __b)) !== false && __a !== nil) {
        return 0
      };
      if ((__a = ((__b = cell['$=='](0)) ? pop['$=='](3) : __b)) !== false && __a !== nil) {
        return 1
      };
      return 0;
    };

    // line 76, (file), GameOfLife#runGeneration
    GameOfLife_prototype.$runGeneration = function() {
      var __a, __b;
      this.genn = (__b = __range(0, this.grid.$x(), false).$to_a(), __b.$map._p = (__a = function(x) {

        var __a, __b;
        if (this.grid == null) this.grid = nil;

        if (x == null) x = nil;

        return (__b = __range(0, this.grid.$y(), false).$to_a(), __b.$map._p = (__a = function(y) {

          
          if (y == null) y = nil;

          return this.$_getState(x, y)
        }, __a._s = this, __a), __b.$map())
      }, __a._s = this, __a), __b.$map());
      this.$_rotateGenerations();
      this.genc = this.genn;
      return this.gen = this.gen['$+'](1);
    };

    // line 87, (file), GameOfLife#_rotateGenerations
    GameOfLife_prototype.$_rotateGenerations = function() {
      var genn = nil;
      if (this.genlog.$length()['$>'](this.history)) {
        this.genlog.$pop()
      };
      genn = this.genn;
      return this.genlog.$unshift(JSON.stringify(genn));
    };

    // line 93, (file), GameOfLife#checkStable
    GameOfLife_prototype.$checkStable = function() {
      var genc = nil, json_genc = nil, __a, __b;
      genc = this.genc;
      json_genc = JSON.stringify(genc);
      (__b = this.genlog, __b.$each._p = (__a = function(key, gen) {

        var __a, __b;
        if (key == null) key = nil;
if (gen == null) gen = nil;

        if ((__a = ((__b = gen['$=='](json_genc)) ? key['$>'](0) : __b)) !== false && __a !== nil) {
          return true
          } else {
          return nil
        }
      }, __a._s = this, __a), __b.$each());
      return false;
    };

    // line 102, (file), GameOfLife#_sizeCanvas
    GameOfLife_prototype.$_sizeCanvas = function() {
      var canvas = nil, width = nil, height = nil, __a, __b;
      canvas = this.canvas;
      width = (__a = this.grid.$x(), __b = this.cell.$w(), typeof(__a) === 'number' ? __a * __b : __a['$*'](__b));
      height = (__a = this.grid.$y(), __b = this.cell.$h(), typeof(__a) === 'number' ? __a * __b : __a['$*'](__b));
      canvas.canvas.width = width;
      canvas.canvas.height = height ;
      return true;
    };

    // line 114, (file), GameOfLife#render
    GameOfLife_prototype.$render = function() {
      var canvas = nil, genc = nil, cell = nil, __a, __b;
      this.$_sizeCanvas();
      canvas = this.canvas;
      genc = this.genc;
      cell = this.cell;
      (__b = __range(0, this.grid.$x(), false).$to_a(), __b.$each._p = (__a = function(x) {

        var __a, __b;
        if (this.grid == null) this.grid = nil;

        if (x == null) x = nil;

        return (__b = __range(0, this.grid.$y(), false).$to_a(), __b.$each._p = (__a = function(y) {

          
          if (y == null) y = nil;

          canvas.fillStyle = genc[x][y] ? "black" : "white";
          return canvas.fillRect(x * cell.w, y * cell.h, cell.w, cell.h);
        }, __a._s = this, __a), __b.$each())
      }, __a._s = this, __a), __b.$each());
      return true;
    };

  })(self, null);
  (function(__base, __super){
    // line 135, (file), class Interval
    function Interval() {};
    Interval = __klass(__base, __super, "Interval", Interval);
    var Interval_prototype = Interval.prototype, __scope = Interval._scope, TMP_1;
    Interval_prototype.interval = nil;

    // line 136, (file), Interval#initialize
    Interval_prototype.$initialize = TMP_1 = function(time) {
      var __context, block;
      block = TMP_1._p || nil, __context = block._s, TMP_1._p = null;
      if (time == null) {
        time = 0
      }
      return this.interval = setInterval(function(){block.$call()}, time);
    };

    // line 140, (file), Interval#stop
    Interval_prototype.$stop = function() {
      var interval = nil;
      interval = this.interval;
      return clearInterval(interval);
    };

  })(self, null);
  (function(__base, __super){
    // line 146, (file), class DoIt
    function DoIt() {};
    DoIt = __klass(__base, __super, "DoIt", DoIt);
    var DoIt_prototype = DoIt.prototype, __scope = DoIt._scope, TMP_2;
    DoIt_prototype.times = DoIt_prototype.interval = nil;

    // line 148, (file), DoIt#initialize
    DoIt_prototype.$initialize = TMP_2 = function(time) {
      var myself = nil, __a, __b, __context, block;
      block = TMP_2._p || nil, __context = block._s, TMP_2._p = null;
      
      this.times = 0;
      myself = this;
      return this.interval = (__b = __scope.Interval, __b.$new._p = (__a = function() {

        
        
        return myself.$work()
      }, __a._s = this, __a), __b.$new(2000));
    };

    // line 156, (file), DoIt#work
    DoIt_prototype.$work = function() {
      
      this.times = this.times['$+'](1);
      this.$alert(this.times);
      if (this.times['$=='](5)) {
        return this.interval.$stop()
        } else {
        return nil
      };
    };

  })(self, null);
  (function(__base, __super){
    // line 163, (file), class RunIt
    function RunIt() {};
    RunIt = __klass(__base, __super, "RunIt", RunIt);
    var RunIt_prototype = RunIt.prototype, __scope = RunIt._scope;
    RunIt_prototype.game = RunIt_prototype.interval = nil;

    // line 164, (file), RunIt#initialize
    RunIt_prototype.$initialize = function(game, interval) {
      var myself = nil, __a, __b;
      this.game = game;
      myself = this;
      return this.interval = (__b = __scope.Interval, __b.$new._p = (__a = function() {

        
        
        return myself.$work()
      }, __a._s = this, __a), __b.$new(interval));
    };

    // line 172, (file), RunIt#work
    RunIt_prototype.$work = function() {
      var game = nil, __a;
      game = this.game;
      game.$runGeneration();
      game.$render();
      document.getElementById('info').innerHTML = "Generation: " + game.gen;
      if ((__a = game.$checkStable()) !== false && __a !== nil) {
        return this.interval.$stop()
        } else {
        return nil
      };
    };

  })(self, null);
  game = __scope.GameOfLife.$new(document.getElementById('board').getContext("2d"));
  do_it = __scope.RunIt.$new(game, 10);
  return game.$render();
})();
